### DISCUSSION TITLE
How to Create query filter to hide multiple products by ID

### DISCUSSION THREAD

**Discussion by gabriellslvr** [Smile-SA/elasticsuite#3342] [Q&A]
Hi guys.
I'm working with a module from Amasty, that has some Catalog rules by customer groups. One of the rules is to hide some products from non-logged-in customers.

This works perfectly on the catalog/search page, but in the elasticsearch query doesn't... I want to build a preference in the `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::prepareRequest()` to include a filter in the elastic search query to not search by the products id's that the Amasty module should hide.

However, I don't know how to build this query and change the request to get the right ID's from the elastic search excluding the ones from the Amasty module rule.

Can you guys help me to apply this filter in the query before sending the request to the Elasticsearch server?

**Comment by rbayet** [comment]
Hello @gabriellslvr,

Your best approach is probably to combine 
* the ability to define [request-wide filters](https://github.com/Smile-SA/elasticsuite/blob/2.11.x/src/module-elasticsuite-catalog/etc/elasticsuite_search_request.xml#L22) 
* adding a dedicated nested structure similar to those for the prices with an "hidden" / "is_blacklisted" boolean + customer_group_id field there
  * or hijacking the existing [nested price structure](https://github.com/Smile-SA/elasticsuite/blob/affec096d16194addea273dfa35145a42da4213a/src/module-elasticsuite-catalog/etc/elasticsuite_indices.xml#L45) and add that "hidden" / "is_blacklisted" boolean field there
* add the data for that is required for the nested structure and the filter to be applied at indexing time through a datasource that will use the DB data of your Amasty module
  * or extending the ["prices" datasource](https://github.com/Smile-SA/elasticsuite/blob/affec096d16194addea273dfa35145a42da4213a/src/module-elasticsuite-catalog/etc/di.xml#L344) to add that "hidden / is_blacklisted" info 

To build the request filter, you can take inspiration from both the "[search blacklist filter](https://github.com/Smile-SA/elasticsuite/blob/affec096d16194addea273dfa35145a42da4213a/src/module-elasticsuite-catalog/Model/Product/Search/Request/Container/Filter/SearchBlacklist.php#L89)" and the way [the price navigation filter is applied](https://github.com/Smile-SA/elasticsuite/blob/affec096d16194addea273dfa35145a42da4213a/src/module-elasticsuite-catalog/Model/Layer/Filter/Price.php#L153) (since it's also customer-group based).

Cheers
